caffe2/contrib/tensorrt/tensorrt_tranformer.cc

```diff
@@ -0,0 +1,423 @@
+      boundary_inputs.emplace(output);
+    }
+  }
+
+  std::unordered_set<std::string> total_inputs;
+  std::vector<std::string> total_inputs_vec;
+  for(int i = start; i < end; ++i) {
+    const auto& op = pred_net.op(i);
+    for (const auto& input: op.input()) {
+      auto rt = total_inputs.emplace(input);
+      if (rt.second) {
+        if (weights.count(input)) {
+          // We add weights as inputs too
+          total_inputs_vec.emplace_back(input);
+          initialization_list->emplace(input);
+        } else if (boundary_inputs.count(input)) {
+          LOG(INFO) << "Adding boundary input: " << input;
+          total_inputs_vec.emplace_back(input);
+        }
+      }
+    }
+  }
+  return total_inputs_vec;
+}
+
+std::vector<std::string>
+FigureOutputs(const NetDef& pred_net, int start, int end) {
+  std::unordered_set<std::string> all_outputs;
+  std::vector<std::string> all_outputs_vec;
+  std::unordered_set<std::string> ext_outputs;
+  for (const auto& e: pred_net.external_output()) {
+    ext_outputs.emplace(e);
+  }
+  for (int i = start; i < end; ++i) {
+    const auto& op = pred_net.op(i);
+    for (const auto& output: op.output()) {
+      all_outputs.emplace(output);
+      all_outputs_vec.emplace_back(output);
+    }
+  }
+  std::unordered_set<std::string> referred_inputs;
+  for (int i = end; i < pred_net.op_size(); ++i) {
+    const auto& op = pred_net.op(i);
+    for (const auto& input: op.input()) {
+      referred_inputs.emplace(input);
+    }
+  }
+  // Remove the output that is
+  // 1. Not referred by the subsequent ops
+  // 2. Not in the external ouput of the net
+  all_outputs_vec.erase(
+      std::remove_if(
+          all_outputs_vec.begin(),
+          all_outputs_vec.end(),
+          [&ext_outputs, &referred_inputs](const std::string& output) {
+            return (
+                !referred_inputs.count(output) && !ext_outputs.count(output));
+          }),
+      all_outputs_vec.end());
+  return all_outputs_vec;
+}
+
+std::vector<::ONNX_NAMESPACE::ValueInfoProto> ConvertToValueInfo(
+    const std::vector<std::string>& names,
+    const std::unordered_map<std::string, TensorShape>& shape_hints) {
+  std::vector<::ONNX_NAMESPACE::ValueInfoProto> r;
+  for (const auto& s: names) {
+    r.emplace_back();
+    auto& value_info = r.back();
+    value_info.set_name(s);
+    const auto it = shape_hints.find(s);
+    if (it == shape_hints.end()) {
+      LOG(WARNING) << "Cannot get shape of " << s;
+    } else {
+      auto* tensor_type = value_info.mutable_type()->mutable_tensor_type();
+      tensor_type->set_elem_type(
+          ::ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT);
+      auto* shape = tensor_type->mutable_shape();
+      for (int i = 0 ; i < it->second.dims().size(); ++i) {
+        auto* dim = shape->add_dim();
+        dim->set_dim_value(it->second.dims(i));
+      }
+    }
+  }
+  return r;
+}
+
+void PruneWeights(const NetDef& pred_net, NetDef* init_net) {
```

Comment:
A different name, e.g. `FindUsedWeights` would be better. Pruning of weights is already used in a different context (model compression)

Reply:
OK